cores/xmc/Pulse.cpp

```diff
@@ -0,0 +1,101 @@
+#include "Arduino.h"
+
+// Ensure we do not wait here forever by setting maximum and default
+// timeouts for 3 minutes and 1 second (in micro-seconds)
+// Default used if timeout is set at < 6 microseconds
+#define MAX_PULSE_TIMEOUT (3 * 60 * 1000 * 1000)
+#define DEF_PULSE_TIMEOUT (1000 * 1000)
+
+uint32_t pulse_state;       // State to measure pin input HIGH or LOW
+uint32_t mask;              // bit mask for relevant pin
+XMC_GPIO_PORT_t *pulsePort; // Pointer to GPIO port register structure
+uint32_t start_time;        // start time of measured pulse
+uint32_t end_time;          // start time of measured pulse
+
+//****************************************************************************
+// @Local Functions
+//****************************************************************************
+
+/* Measures the length (in microseconds) of a pulse on the Arduino pin;
+ * Works on pulses from 6 microseconds to 3 minutes in length,
+ * but must be called at least a few dozen microseconds
+ * before the start of the pulse.
+ *
+ * Waits for input signal to be at opposite level to being measured
+ * to avoid getting short measurements of wrong pulse.
+ *
+ * Parameters
+ *          pin     - which pin to monitor
+ *          state   - HIGH or LOW, for the level of pulse to measure.
+ *          timeout - max time in microseconds to wait for whole process
+ *
+ * Returns  0 for timeout reached or INVALID pin number
+ *
+ * ATTENTION:
+ * This function relies on the SysTick timer in function micros() so is
+ * LESS prone to interrupts or other activity interfering. However there
+ * will always be the possibility of at LEAST +/- 1 counts difference to
+ * actual signal time. If you need high accuracy use another solution
+ * with hardware timers in capture mode.
+ *
+ * Use of SysTick hardware timer and micros( ) function makes the function
+ * more portable across various micros and CPU clock frequencies.
+ *
+ * Assumptions]
+ *      As per AVR Arduino PulseIn pin is expected to be in input mode already
+ *      No major interrupt activity while measurement is taken (including Serial)
+ *
+ *  Do NOT make changes to use local (stack) variables as this WILL ADD overhead
+ *  to any operations.
+ *  To improve accuracy beyond this will probably need hard coded assembler
+ */
+unsigned long pulseIn(pin_size_t pin, uint8_t state, unsigned long timeout) {
+// check pin is valid by NUM_DIGITAL_PINS, return 0 if invalid
+#ifdef NUM_DIGITAL_PINS
+    if (pin >= NUM_DIGITAL_PINS)
```

Comment:
Please use braces for single statement blocks. If this is not enabled in the formatting tool, we will do soon.